pkg/hub/client/policy_merger.go

```diff
@@ -30,10 +30,28 @@ func (e *PolicyEnforcedClient) mergePolicies() {
 	e.mergedPolicy = currentPolicy
 }
 
-// from new policy we are checking if there are the same rules. If yes we fill missing data,
-// if not we add a rule to the end
-// current policy is a higher priority policy
+// RequiredTypeInstancesForRule returns the merged list of TypeInstances from Rule and Defaults.
+func (e *PolicyEnforcedClient) RequiredTypeInstancesForRule(policyRule policy.Rule) []policy.RequiredTypeInstanceToInject {
+	// prefer policy Rule over Default
+	return mergeRequiredTypeInstances(e.mergedPolicy.Interface.DefaultRequiredTypeInstancesToInject(), policyRule.RequiredTypeInstancesToInject())
+}
+
 func applyInterfacePolicy(currentPolicy *policy.InterfacePolicy, newPolicy policy.InterfacePolicy) {
+	// Default
+	if len(newPolicy.DefaultRequiredTypeInstancesToInject()) > 0 {
+		if currentPolicy.Default == nil || currentPolicy.Default.Inject == nil {
+			currentPolicy.Default = &policy.InterfaceDefault{
+				Inject: &policy.DefaultInject{
+					RequiredTypeInstances: []policy.RequiredTypeInstanceToInject{},
+				},
+			}
+		}
+		currentPolicy.Default.Inject.RequiredTypeInstances = mergeRequiredTypeInstances(currentPolicy.Default.Inject.RequiredTypeInstances, newPolicy.Default.Inject.RequiredTypeInstances)
```

Comment:
I went through the "Backend testing" scenario with one difference - I used the policy not as a global one, but as an Action one.
Unfortunately I got panic on Engine:

```
2022-03-03T18:26:10.804Z	INFO	policy-svc	policy/service.go:82	Getting Policy	{"module": "policyService"}
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x1812a26]

goroutine 384 [running]:
capact.io/capact/pkg/hub/client.mergeRequiredTypeInstances(0xc000b985a0, 0x2, 0x2, 0xc0001f8a00, 0x1, 0x4, 0x1, 0x2, 0x2)
	/capact.io/capact/pkg/hub/client/policy_merger.go:139 +0x126
capact.io/capact/pkg/hub/client.applyInterfacePolicy(0xc000639418, 0xc000592158, 0xc0002e5b00, 0x2, 0x4)
	/capact.io/capact/pkg/hub/client/policy_merger.go:49 +0xd0
capact.io/capact/pkg/hub/client.(*PolicyEnforcedClient).mergePolicies(0xc000968d80)
	/capact.io/capact/pkg/hub/client/policy_merger.go:17 +0x185
capact.io/capact/pkg/hub/client.(*PolicyEnforcedClient).SetActionPolicy(0xc000968d80, 0xc0002822d0, 0xc000166240, 0x2, 0x4, 0x2d2b9d8, 0x0, 0x0)
	/capact.io/capact/pkg/hub/client/policy_enforced_client.go:271 +0x78
capact.io/capact/pkg/sdk/renderer/argo.WithActionPolicy.func1(0xc000a70680)
	/capact.io/capact/pkg/sdk/renderer/argo/options.go:36 +0x83
capact.io/capact/pkg/sdk/renderer/argo.newDedicatedRenderer(0xc000a78a80, 0x32, 0x20779c8, 0xc000968d80, 0xc000a1d380, 0xc000b98500, 0x6, 0xa, 0x40e118)
	/capact.io/capact/pkg/sdk/renderer/argo/dedicated_renderer.go:76 +0x14e
capact.io/capact/pkg/sdk/renderer/argo.(*Renderer).Render(0xc000a1d3c0, 0x205f4c8, 0xc00079c6c0, 0xc000b98550, 0x0, 0x0, 0x0)
	/capact.io/capact/pkg/sdk/renderer/argo/renderer.go:79 +0x2bb
capact.io/capact/internal/k8s-engine/controller.(*ActionService).RenderAction(0xc000a61b80, 0x205f4c8, 0xc00079c6c0, 0xc00000af00, 0x0, 0x0, 0x0)
	/capact.io/capact/internal/k8s-engine/controller/action_service.go:411 +0x639
capact.io/capact/internal/k8s-engine/controller.(*ActionReconciler).renderAction(0xc000a78f00, 0x205f4c8, 0xc00079c6c0, 0xc00000af00, 0x0, 0x0, 0xe, 0x2087560)
	/capact.io/capact/internal/k8s-engine/controller/action_controller.go:197 +0x73
capact.io/capact/internal/k8s-engine/controller.(*ActionReconciler).Reconcile(0xc000a78f00, 0x205f4c8, 0xc00079c6c0, 0xc00088a809, 0x7, 0xc00088a7f0, 0xe, 0xc00079c6c0, 0xc00003a000, 0x1c3a3e0, ...)
	/capact.io/capact/internal/k8s-engine/controller/action_controller.go:137 +0x905
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000a61c20, 0x205f420, 0xc000c15880, 0x1bd7d40, 0xc000a7c200)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:298 +0x30d
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000a61c20, 0x205f420, 0xc000c15880, 0x0)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:253 +0x205
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2(0xc0004f4030, 0xc000a61c20, 0x205f420, 0xc000c15880)
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.9.6/pkg/internal/controller/controller.go:214 +0x6b
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2
```

Looks like this line is problematic? 🤔 
```
if newTI.TypeRef.Path == ti.TypeRef.Path && newTI.TypeRef.Revision == ti.TypeRef.Revision {
```

which would mean that the TypeRefs weren't resolved after all? Probably because merging policies happen during `SetActionPolicy`:

```go
// SetActionPolicy sets policy to use during actiom workflow rendering. This setter is thread safe.
func (e *PolicyEnforcedClient) SetActionPolicy(p policy.ActionPolicy) {
	e.mu.Lock()
	e.actionPolicy = policy.Policy(p)
	e.mergePolicies()
	e.mu.Unlock()
}
```

Please take a look on that, I didn't spend much time on it, but it needs to be fixed.

Also, in the line I marked, let's check whether the newPolicy.Default and Inject are not nil, to avoid panic in this place if defaults are not used in one of the policies 🙂 
```go
currentPolicy.Default.Inject.RequiredTypeInstances = mergeRequiredTypeInstances(currentPolicy.Default.Inject.RequiredTypeInstances, newPolicy.Default.Inject.RequiredTypeInstances)
```

---

What I did:

```bash
capact act create cap.interface.capactio.capact.validation.action.passing --name inject-storage --type-instances-from-file /tmp/act-input-ti.yaml --action-policy-from-file=/tmp/inject-storage-policy.yaml
```

Action Policy:
```yaml
interface:
  default:
    inject:
        requiredTypeInstances:
        - description: Test TypeInstance
          id: 1ef12f6c-5cae-43a0-b3d5-fe851d61adf3
        - description: Helm backend TypeInstance
          id: 48d6ef24-43c4-4604-96c2-057bf4e37abf
  rules:
  - interface:
      path: cap.interface.capactio.capact.validation.action.passing
    oneOf:
    - implementationConstraints:
        attributes:
        - path: cap.attribute.capactio.capact.validation.policy.most-preferred
        requires:
        - path: cap.type.capactio.capact.validation.single-key
    - implementationConstraints:
        path: cap.implementation.capactio.capact.validation.action.passing-a
  - interface:
      path: cap.*
    oneOf:
    - implementationConstraints: {}
typeInstance:
  rules: []
```

Global:
```yaml
   interface:
      default:
        inject:
          requiredTypeInstances:
          - description: AWS credentials
            id: e8c77271-8129-4c30-85b4-5c20ef78cc44
      rules:
      - interface:
          path: cap.interface.database.postgresql.install
          revision: null
        oneOf:
        - implementationConstraints:
            attributes:
            - path: cap.attribute.cloud.provider.aws
              revision: null
      - interface:
          path: cap.*
          revision: null
        oneOf:
        - implementationConstraints:
            requires:
            - path: cap.core.type.platform.kubernetes
              revision: null
        - implementationConstraints: {}
    typeInstance:
      rules: null
```

Reply:
Yes, as discussed f2f it was connected to the merging of the policy. Should be fine now. :)